src/storage/btree_load.c

```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
     }
   return PERF_PAGE_BTREE_ROOT;
 }
+
+/*
+ * btree_load_check_fk () -
+ *
+ *    return:				- NO_ERROR or error code.
+ *    load_args_local(in):		- Information about the foreign key
+ *    sort_args_local(in):		- Information about the primary key
+ *
+ */
+int
+btree_load_check_fk (THREAD_ENTRY * thread_p, const LOAD_ARGS * load_args, const SORT_ARGS * sort_args)
+{
+  DB_VALUE fk_key, pk_key;
+  int fk_key_cnt = -1, pk_key_cnt = -1;
+  BTREE_NODE_HEADER *fk_header = NULL, *pk_header = NULL;
+  VPID vpid;
+  int ret = NO_ERROR;
+  PAGE_PTR curr_fk_pageptr = NULL, old_page = NULL;
+  INDX_SCAN_ID pk_isid;
+  BTREE_SCAN pk_bt_scan;
+  INT16 fk_slot_id = -1;
+  bool found = false;
+  int advance_direction = 0;
+  char *val_print = NULL;
+  bool is_fk_scan_desc = false;
+  MVCC_SNAPSHOT mvcc_snapshot_dirty;
+  int lock_ret = LK_GRANTED;
+  DB_VALUE_COMPARE_RESULT compare_ret;
+
+  DB_MAKE_NULL (&fk_key);
+  DB_MAKE_NULL (&pk_key);
+  mvcc_snapshot_dirty.snapshot_fnc = mvcc_satisfies_dirty;
+
+  /* Initialize index scan on primary key btid. */
+  scan_init_index_scan (&pk_isid, NULL, NULL);
+  BTREE_INIT_SCAN (&pk_bt_scan);
+  pk_isid.copy_buf_len = DBVAL_BUFSIZE;
+  pk_isid.check_not_vacuumed = true;
+
+  /* Lock the primary key class. */
+  lock_ret = lock_object (thread_p, sort_args->fk_refcls_oid, oid_Root_class_oid, SIX_LOCK, LK_COND_LOCK);
+  if (lock_ret != LK_GRANTED)
+    {
+      ASSERT_ERROR_AND_SET (ret);
+      return ret;
+    }
+
+  /* Primary key index search prepare */
+  ret = btree_prepare_bts (thread_p, &pk_bt_scan, sort_args->fk_refcls_pk_btid, &pk_isid, NULL, NULL, NULL,
+			   NULL, NULL, false, NULL);
+  if (ret != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      return ret;
+    }
+
+  /* Set the order. */
+  is_fk_scan_desc = sort_args->key_type->is_desc != pk_bt_scan.btid_int.key_type->is_desc;
+
+  /* Get the corresponding leaf of the foreign key. */
+  if (!is_fk_scan_desc)
+    {
+      /* Get first leaf vpid. */
+      vpid = load_args->vpid_first_leaf;
+    }
+  else
+    {
+      /* Get the last leaf. Current leaf is the last one. */
+      vpid = load_args->leaf.vpid;
+    }
+
+  /* Init slot id */
+  pk_bt_scan.slot_id = 0;
+
+  while (true)
+    {
+      ret = btree_advance_to_next_slot_and_fix_page (thread_p, sort_args->btid, &vpid, &curr_fk_pageptr, &fk_slot_id,
+						     &fk_key, sort_args->key_type->is_desc, &fk_key_cnt, &fk_header,
+						     &mvcc_snapshot_dirty);
```

Comment:
I think I told you to add mvcc_snapshot_dirty here, but I was wrong. btree_sort_get_next already checks dirty snapshot, therefore only objects passing the checks are loaded. all keys will pass the visibility check in foreign key.

However, the dirty snapshot is required to check primary keys.